### PR TITLE
User-service implementation v1.0

### PR DESCRIPTION
**Pull request by Valelaaa**
- first implementation of UserService

### CODE REVIEW

File: src/main/java/com/endava/internship/service/UserServiceImpl.java [hunk 1]
```diff
@@ -2,68 +2,122 @@
...
+                .distinct().
+                collect(toList());
     }
 
     @Override
     public Optional<User> getUpdateUserWithAgeHigherThan(final List<User> users, final int age) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
+                .filter(user -> user.getAge() > age && user.getPrivileges()
+                        .contains(Privilege.UPDATE)).findFirst();
```
Comment (NataliaIovva) [inline src/main/java/com/endava/internship/service/UserServiceImpl.java]: please `user.getPrivileges().contains(Privilege.UPDATE)` in one line and maybe add "()" to easier reading code 
and static import for UPDATE

File: src/main/java/com/endava/internship/service/UserServiceImpl.java [hunk 2]
```diff
@@ -2,68 +2,122 @@
...
+        return users.stream()
+                .map(user -> Pair.of(user.getPrivileges().size(), user))
+                .collect(groupingBy(Pair::getKey, mapping(Pair::getValue, toList())));
     }
 
     @Override
-    public double getAverageAgeForUsers(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+    public Optional<Double> getAverageAgeForUsers(final List<User> users) {
+        DoubleSummaryStatistics doubleSummaryStatistics = users.stream()
```
Comment (NataliaIovva) [inline src/main/java/com/endava/internship/service/UserServiceImpl.java]: `final DoubleSummaryStatistics doubleSummaryStatistics = users.stream().collect(summarizingDouble(User::getAge));`

File: src/main/java/com/endava/internship/service/UserServiceImpl.java [hunk 3]
```diff
@@ -2,68 +2,122 @@
...
+                        .thenComparing(User::getFirstName))
+                .collect(toList());
     }
 
     @Override
     public List<Privilege> getAllDistinctPrivileges(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
+                .flatMap(user -> user.getPrivileges().stream())
+                .distinct().
```
Comment (NataliaIovva) [inline src/main/java/com/endava/internship/service/UserServiceImpl.java]: `.distinct()
.collect(toList());`

File: src/test/java/com/endava/internship/service/UserServiceImplTest.java [hunk 4]
```diff
@@ -98,23 +98,22 @@ void shouldReturnGroupedMapByNumberOfPrivileges() {
 
     @Test
     void shouldReturnAverageAgeForUsers() {
-        final int expectedAverage = 23;
+        Double value = Double.valueOf(23);
+        final Optional<Double> expectedAverage = Optional.of(value);
```
Comment (NataliaIovva) [inline src/test/java/com/endava/internship/service/UserServiceImplTest.java]: `final Optional<Double> expectedAverage = Optional.of(23.0);`

File: src/test/java/com/endava/internship/service/UserServiceImplTest.java [hunk 5]
```diff
@@ -131,13 +130,40 @@ void shouldReturnMostFrequentLastName() {
         assertThat(mostFrequentLastName)
                 .hasValueSatisfying(name -> assertThat(name).isEqualToIgnoringCase(mostFrequentName));
     }
+    @Test
+    void shouldReturnOptionalEmptyIfThereAreNoDistinctNumberOfLastNames(){
+        final String mostFrequentName = "Smith";
```
Comment (NataliaIovva) [inline src/test/java/com/endava/internship/service/UserServiceImplTest.java]: remove unused variable

File: src/test/java/com/endava/internship/service/UserServiceImplTest.java [hunk 6]
```diff
@@ -131,13 +130,40 @@ void shouldReturnMostFrequentLastName() {
...
+        assertThat(mostFrequentLastName).isEmpty();
+    }
+    @Test
+    void shouldReturnTheOnlyOneUserName() {
+        final String mostFrequentName = "Doe";
 
+        final User user1 = new User(1L, "John", "Doe", 26, singletonList(Privilege.UPDATE));
+
+        final Optional<String> mostFrequentLastName =
+                userService.getMostFrequentLastName(asList(user1));
```
Comment (NataliaIovva) [inline src/test/java/com/endava/internship/service/UserServiceImplTest.java]: use singletonList
```
final Optional<String> mostFrequentLastName =
              userService.getMostFrequentLastName(singletonList(user1));
```

File: src/main/java/com/endava/internship/service/UserServiceImpl.java (hunk 1 above)
Comment (Valelaaa) [reply src/main/java/com/endava/internship/service/UserServiceImpl.java]: Done

File: src/main/java/com/endava/internship/service/UserServiceImpl.java (hunk 2 above)
Comment (Valelaaa) [reply src/main/java/com/endava/internship/service/UserServiceImpl.java]: Fixed

File: src/main/java/com/endava/internship/service/UserServiceImpl.java (hunk 3 above)
Comment (Valelaaa) [reply src/main/java/com/endava/internship/service/UserServiceImpl.java]: fixed

File: src/test/java/com/endava/internship/service/UserServiceImplTest.java (hunk 4 above)
Comment (Valelaaa) [reply src/test/java/com/endava/internship/service/UserServiceImplTest.java]: done

File: src/test/java/com/endava/internship/service/UserServiceImplTest.java (hunk 5 above)
Comment (Valelaaa) [reply src/test/java/com/endava/internship/service/UserServiceImplTest.java]: done!

File: src/test/java/com/endava/internship/service/UserServiceImplTest.java (hunk 6 above)
Comment (Valelaaa) [reply src/test/java/com/endava/internship/service/UserServiceImplTest.java]: fixed!

**Review by NataliaIovva — APPROVED** [review]

**Pull request merged**